src/main/java/org/apache/commons/io/input/QueueInputStream.java

```diff
@@ -224,4 +226,49 @@ public int read() {
         }
     }
 
+    /**
+     * Reads up to {@code length} bytes of data from the input stream into
+     * an array of bytes.  The first byte is read while honoring the timeout; the rest are read while <i>not</i> honoring
+     * the timeout. The number of bytes actually read is returned as an integer.
+     *
+     * @param b     the buffer into which the data is read.
+     * @param offset   the start offset in array {@code b} at which the data is written.
+     * @param length   the maximum number of bytes to read.
+     * @return     the total number of bytes read into the buffer, or {@code -1} if there is no more data because the
+     *              end of the stream has been reached.
+     * @throws NullPointerException If {@code b} is {@code null}.
+     * @throws IllegalStateException if thread is interrupted while waiting for the first byte.
+     * @throws IndexOutOfBoundsException if {@code offset} is negative, {@code length} is negative, or {@code length} is
+     *             greater than {@code b.length - offset}.
+     * @since 2.20.0
+     */
+    @Override
+    public int read(final byte[] b, final int offset, final int length) {
+        if (b == null) {
+            throw new NullPointerException();
+        } else if (offset < 0 || length < 0 || length > b.length - offset) {
+            throw new IndexOutOfBoundsException(
+                    String.format("Range [%d, %<d + %d) out of bounds for length %d", offset, length, b.length));
+        } else if (length == 0) {
+            return 0;
+        }
+        final List<Integer> drain = new ArrayList<>(Math.min(length, blockingQueue.size()));
+        blockingQueue.drainTo(drain, length);
+        if (drain.isEmpty()) {
+            // no data immediately available. wait for first byte
+            final int value = read();
+            if (value == EOF) {
+                return EOF;
+            }
+            drain.add(value);
+            blockingQueue.drainTo(drain, length - 1);
```

Comment:
I think this call to `drainTo(...)` could be wrong because it can cause EOF markers to end up in the `drain` data.
We want to drain until we get the requested length _or_ the EOF marker is seen in the drained data.
For example:
- Thread A is paused just before, on, or during the `drain.add(value)` above, but before `blockingQueue.drainTo(drain, length - 1);`
- Thread B adds a `-1` (an EOF marker).
- Thread A resumes, calls `blockingQueue.drainTo(drain, length - 1);` and now contains the EOF marker in `drain` and includes it in `b`.

Reply:
The queue itself does not contain any `-1` elements:

https://github.com/apache/commons-io/blob/5080fa310dc19c20f9631bd18cfc3a5a63ef160e/src/main/java/org/apache/commons/io/output/QueueOutputStream.java#L92-L101

The `EOF` value that appears on lines 260 and 261 is generated by the `read()` method to indicate a missing value.

Reply:
That's an interesting question on whether to specially treat -1 in the underlying data stream. Looking at the behavior of PipedInputStream and ByteArrayInputStream, no special meaning is attached to -1. If there is a -1 and if there is data subsequently available, all data is returned as is.

I think QueueInputStream should behave similarly.